### PR TITLE
openssl critical sections

### PR DESCRIPTION
**Pull request by AlexK0**
Call openssl methods under signal critical sections.

### CODE REVIEW

File: runtime/critical_section.h [hunk 1]
```diff
@@ -21,6 +21,25 @@ struct CriticalSectionGuard : private vk::not_copyable {
   ~CriticalSectionGuard() noexcept { leave_critical_section(); }
 };
 
+class CriticalSectionSmartGuard : private vk::not_copyable {
+public:
+  CriticalSectionSmartGuard() noexcept { enter_critical_section(); }
+
+  void leave_critical_section() noexcept {
```
Comment (DrDet) [inline runtime/critical_section.h:28]: Maybe it's worth to add similar `enter_critical_section` method to make it more smart :) ? Having only one of them looks a bit strange

Comment (AlexK0) [reply runtime/critical_section.h:28]: It was added in the next commit

File: runtime/critical_section.h [hunk 2]
```diff
@@ -21,6 +21,30 @@ struct CriticalSectionGuard : private vk::not_copyable {
...
+    in_critical_section_ = false;
+  }
+
+  void enter_critical_section() noexcept {
+    dl::enter_critical_section();
+    in_critical_section_ = true;
+  }
+
+  ~CriticalSectionSmartGuard() noexcept {
+    if (in_critical_section_) {
```
Comment (eugene536) [inline runtime/critical_section.h:39]: maybe put this if inside the leave_critical_section method?

File: runtime/critical_section.h [hunk 3]
```diff
@@ -26,13 +26,17 @@ class CriticalSectionSmartGuard : private vk::not_copyable {
...
   void enter_critical_section() noexcept {
-    dl::enter_critical_section();
-    in_critical_section_ = true;
+    if (!in_critical_section_) {
+      dl::enter_critical_section();
+      in_critical_section_ = true;
+    }
   }
 
   ~CriticalSectionSmartGuard() noexcept {
```
Comment (eugene536) [inline runtime/critical_section.h:42]: and now you can just call the `leave_critical_section` method here, without `if`

**Review by eugene536 — APPROVED** [review]

**Pull request merged**